prow/cmd/branchprotector/protect_test.go

```diff
@@ -697,6 +707,61 @@ branch-protection:
 				},
 			},
 		},
+		{
+			name:     "do not make update request if the branch is already up-to-date",
+			branches: []string{"kubernetes/test-infra=master"},
+			config: `
+branch-protection:
+  enforce_admins: true
+  required_status_checks:
+    contexts:
+    - config-presubmit
+    strict: true
+  required_pull_request_reviews:
+    required_approving_review_count: 3
+    dismiss_stale: false
+    require_code_owner_reviews: true
+    dismissal_restrictions:
+      users:
+      - bob
+      - jane
+      teams:
+      - oncall
+      - sres
+  restrictions:
+    teams:
+    - config-team
+    users:
+    - cindy
+  protect: true
+  orgs:
+    kubernetes:
+      repos:
+        test-infra:
+`,
+			branchProtections: map[string]github.BranchProtection{
+				"kubernetes/test-infra=master": {
+					EnforceAdmins: github.EnforceAdmins{Enabled: true},
+					RequiredStatusChecks: &github.RequiredStatusChecks{
+						Strict:   true,
+						Contexts: []string{"config-presubmit"},
+					},
+					RequiredPullRequestReviews: &github.RequiredPullRequestReviews{
+						DismissStaleReviews:          false,
+						RequireCodeOwnerReviews:      true,
+						RequiredApprovingReviewCount: 3,
+						DismissalRestrictions: github.Restrictions{
+							Users: &[]string{"bob", "jane"},
+							Teams: &[]string{"oncall", "sres"},
+						},
+					},
+					Restrictions: &github.Restrictions{
+						Users: &[]string{"cindy"},
+						Teams: &[]string{"config-team"},
+					},
+				},
+			},
+		},
```

Comment:
It would be nice to have a test case for "make request if branch protection is present, but out of date" and a test case for "don't remove branch protection if protection is not present and config sets `protect: false`" (which wont work until the `req == nil && currentBP == nil` case is skipped).

Reply:
Thank you for the review. Added the first test case. Will hold off on the second test case to see if the existing optimisation and unit tests already cover it.